[NEW] Dev Dockerfile silently masks `npm install` failures

## Verification pass — 2026-05-02

`web/Dockerfile:6`:

```dockerfile
RUN npm install || true
```

A failed `npm install` produces a "successful" image with a broken/partial `node_modules`. Every dev container build is a coin flip on package integrity. Symptoms surface later as "module not found" or stale lockfile drift.

## Suggested fix

Drop `|| true`:

```dockerfile
RUN npm ci --prefer-offline
```

If transient registry flakes are the historical motivation, retry instead of swallow:

```dockerfile
RUN for i in 1 2 3; do npm ci --prefer-offline && break || sleep 5; done
```

## Verify

`docker compose -f docker-compose.dev.yml build web` should fail on a busted `package-lock.json`, not succeed silently.

## Implementation plan

### Context

`web/Dockerfile:6`:

```dockerfile
RUN npm install || true
```

`|| true` swallows non-zero exit codes from `npm install`, so a registry
flake, a lockfile inconsistency (npm exits non-zero on `package.json` ↔
`package-lock.json` drift), or a postinstall failure produces a
"successful" image with a partially populated `node_modules`. Symptoms
surface later as `Cannot find module '...'` at Vite startup or as
`tsc -b` errors that disappear on a clean rebuild.

Compose-side this is the dev image (`web` service in
`docker-compose.yml` runs `vite --reload` per CLAUDE.md). Prod doesn't
use this Dockerfile; prod ships the static bundle baked elsewhere.

The dev `web` service mounts the host `web/` directory into `/app` so
in normal day-to-day use `node_modules/` is overwritten by host volume
mounting anyway — but the image itself still gets used as the source
of truth on a fresh clone before the first `npm install` host-side, and
on CI runners that don't pre-populate `node_modules`.

CLAUDE.md flags one related lesson: `web/vite.config.js` is auto-emitted
by the composite TypeScript project and gitignored. Anything that
depends on a working `node_modules` to emit it must actually have a
working `node_modules`.

### Approach

One-line change. Replace `RUN npm install || true` with `npm ci
--prefer-offline` wrapped in a 3-attempt retry loop so transient
registry errors retry instead of silently swallow:

```dockerfile
RUN for i in 1 2 3; do npm ci --prefer-offline && break || sleep 5; done
```

`npm ci` is the right verb for a Dockerfile: it requires the lockfile,
fails on lockfile drift, and is reproducible. `--prefer-offline` keeps
warm-cache builds fast without forbidding network access (which `--offline`
would). The retry loop covers the historical motivation (transient
registry blips during `docker compose up --build`) without masking
genuine package failures: after three real failures, the build exits
non-zero and `docker compose up --build` aborts loudly.

Don't introduce a `.npmrc` — the existing `package-lock.json` already
pins versions and `npm ci` already refuses to mutate it.

### Files

- `web/Dockerfile:6` — single-line replacement. The surrounding
  `COPY package.json package-lock.json* ./` (line 5) and `COPY . .`
  (line 8) stay intact. Glob `package-lock.json*` keeps a missing
  lockfile from breaking the COPY, but `npm ci` requires the lockfile
  so a missing one will now hard-fail; that's the desired behaviour.

### Helpers

None. Pure Dockerfile edit.

### Tests

Manual validation:
1. `docker compose build web` on a clean repo — succeeds, `node_modules/`
   is fully populated.
2. `git checkout web/package-lock.json -- <stale-rev>; docker compose
   build web` — fails with a clear "lockfile out of sync" error
   instead of building a half-broken image.
3. Deliberately break `package.json` (e.g. inject a bogus dep), then
   `docker compose build web` — fails on the bogus dep, retry loop
   tries three times (visible in the build log), then aborts.

No automated test is justified — this is one line of build glue.

### Scope

In: `web/Dockerfile` line 6 only.

Out: `web/package.json`, `web/package-lock.json`, anything in
`docker-compose.yml`/`docker-compose.prod.yml`, the prod web build
pipeline.

### Verification

- `docker compose build web` from a clean clone succeeds.
- `docker compose up --build` brings the dev stack up; `vite` starts
  on port 5173.
- Hot reload still works (the host volume mount in
  `docker-compose.yml` is unchanged).
- A simulated registry failure (e.g. `--build-arg http_proxy=http://0.0.0.0:1`
  to force npm to fail) hits the retry loop three times and then the
  build exits non-zero — *not* zero.

### Notes

- CLAUDE.md hard-invariant flagged: none of the hard invariants are
  touched by this change. The closest related rule is "Things that
  have bitten us": `web/vite.config.js` is auto-emitted by `tsc -b`
  via `vite build`, which depends on a working `node_modules`. Fixing
  this Dockerfile makes that auto-emit reliable too.
- `npm ci` is more demanding than `npm install` about lockfile state;
  contributors who edit `package.json` without running `npm install`
  locally first will see the build fail. That's the intended trade —
  caught at build time instead of as a runtime "module not found".
- Don't switch to `pnpm` or `bun` here. The CI workflow at
  `.github/workflows/ci.yml:85` runs `npm ci`, and changing the
  package manager is a much bigger change than this issue is.
- The historical "registry flake" motivation isn't documented in the
  repo, but the retry loop is a cheap insurance policy. If a year
  passes without the loop ever entering iteration 2, drop it then.

--- a/web/Dockerfile
+++ b/web/Dockerfile
@@ -3,7 +3,7 @@ FROM node:20-alpine
 WORKDIR /app
 
 COPY package.json package-lock.json* ./
-RUN npm install || true
+RUN for i in 1 2 3; do npm ci --prefer-offline && break || sleep 5; done
 
 COPY . .
 

